WLC/code_fixing/codefixer.py

```diff
@@ -20,40 +22,70 @@ def __init__(self, code, indents, poss_lines):
         self.code = code
         self.indents = indents
         self.poss_lines = poss_lines
-        self.context = {'variables': [], 'functions': get_functions(), 'classes': [], 'imports': [], 'methods': []}
-        self.in_class = False
-        self.class_indent = 0
+
         self.syntax = []
         self.rules = []
         self.statements = []
 
+        self.curr_line_n = 0
+        self.context = {'variables': [], 'functions': get_functions(), 'classes': [], 'imports': [], 'methods': []}
+
+        self.class_indent = 0
+        self.class_context = collections.defaultdict(lambda: collections.defaultdict(list))
+        self.curr_class = None
+
+        self.def_indent = 0
+        self.def_context = collections.defaultdict(lambda: collections.defaultdict(list))
+        self.curr_def = None
+
         self.syntax.append(('VARIABLE', '[a-z_]+'))
         self.syntax.append(('FUNCTION', '[a-z_]+'))
 
-        self.syntax.append(('STATEMENT', '.*'))
-        self.syntax.append(('BOOLEAN', '.*'))
+        self.syntax.append(('STATEMENT', '.+?'))  # Note - should be '.+?' or '.+' ?
```

Comment:
There should not be '?' at the end.

Reply:
If you think greedy matching is best, I'll do that. I do think that non-greedy might be a good idea though.